src/Interpreters/Context.cpp

```diff
@@ -1790,7 +1790,7 @@ StoragePtr Context::executeTableFunction(const ASTPtr & table_expression, const
         }
 
         uint64_t use_structure_from_insertion_table_in_table_functions = getSettingsRef().use_structure_from_insertion_table_in_table_functions;
-        if (use_structure_from_insertion_table_in_table_functions && table_function_ptr->needStructureHint() && hasInsertionTable())
+        if (select_query_hint && use_structure_from_insertion_table_in_table_functions && table_function_ptr->needStructureHint() && hasInsertionTable())
```

Comment:
Probably some comment would be helpful.

Reply:
This is new code to me, I need more context to get it straight. One way to simplify the complex condition is to assign it to a variable with a self-explanatory name. Probably, it can help